src/rgw/rgw_admin.cc

```diff
@@ -8286,6 +8353,189 @@ int main(int argc, const char **argv)
     }
   }
 
+  if (opt_cmd == OPT::BUCKET_CHECK_OLH || opt_cmd == OPT::BUCKET_CHECK_UNLINKED) {
+    rgw::sal::RadosStore* rados_store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+    if (!rados_store) {
+      cerr <<
+	      "WARNING: this command can only work when the cluster has a RADOS backing store." <<
+	      std::endl;
+      return 0;
+    }
+    RGWRados* store = rados_store->getRados();
+    if (bucket_name.empty()) {
+      cerr << "ERROR: bucket name not specified" << std::endl;
+      return EINVAL;
+    }
+    ret = init_bucket(user.get(), tenant, bucket_name, bucket_id, &bucket);
+    if (ret < 0) {
+      cerr << "ERROR: could not init bucket: " << cpp_strerror(-ret) << std::endl;
+      return -ret;
+    }
+    if ((bucket->get_info().versioning_status() & BUCKET_VERSIONED) == 0) {
+      cerr << "WARNING: this command is not applicable to unversioned buckets" << std::endl;
+      return 0;
+    }
+
+    if (max_entries < 0) {
+      max_entries = 1000;
+    }
+
+    RGWObjectCtx obj_ctx(driver);
+
+    const auto& index = bucket->get_info().layout.current_index;
+    const int max_shards = rgw::num_shards(index);
+
+    if (fix) {
+      cerr << "NOTICE: fix=true" << std::endl;
+    }
+
+    ceph::real_clock::time_point now = ceph::real_clock::now();
+    ceph::real_clock::time_point not_after = now - std::chrono::hours(min_age);
+
+    std::string verb = fix ? "removed" : "found";
+    int num_threads = std::min(max_shards, max_threads);
+    std::mutex cerr_mutex;
+    atomic_int next_shard = (specified_shard_id ? shard_id : 0);
+    atomic_ullong total_count = 0;
+    std::vector<std::thread> ts;
+    for (int j=0; j<num_threads; j++) {
+      std::thread t([&]() {
+        list<rgw_cls_bi_entry> entries;
+        bool is_truncated;
+        while (true) {
+          int shard = next_shard.fetch_add(1, std::memory_order_relaxed);
+          if (shard >= max_shards) {
+            return;
+          }
+          RGWRados::BucketShard bs(store);
+          string marker = opt_cmd == OPT::BUCKET_CHECK_OLH ? "\x80" "1001_" : "\x80" "1000_";
```

Comment:
Yeah I know. I don't feel great about that either, but it's a significant optimization to skip listing all of the irrelevant parts of the keyspace and I didn't think it would be a great idea to add new cls methods just for these specialty listing cases.